File: hash_tables.py
import random
class HashTable:
    def __init__(self, size: int = 2000):
        self.obj = [[0]*2 for i in range(size)]
     
    def __hash(self, key: str=""):
        if key:
            address = 0
            for i in key:
                random.seed(42)
                address += (address + int((ord(i)/5)) +  random.randint(1, 3)) % len(self.obj)
            
            return address
    
    def set(self, key, value):
        location  = self.__hash(key)
        if self.obj[location] == [0,0]:
            self.obj[location] = [key, value]
        else:
            print("location already exists, and want to avoid collision")
        return self.obj[location]

    def get(self, key):
        location  = self.__hash(key)
        data = self.obj[location]
        if data != [0,0]:
            if isinstance(data[0], str):
                return data[1]
            else:
                index = 0
                for i in data:
                    if i[index][0] == key:
                        return i[index][1]
                    index += 1
        else:
            return None
        
    def keys(self):
        return [i[0] for i in self.obj if i != [0, 0]]

if __name__ == "__main__":
    ht = HashTable()
    ht.set("Nakul", "yo")
    ht.set("Evelyn", "why")
    ht.set("Amigo", "weon")
    ht.set("Amigo", "weon")
    print(ht.get("Amigo"))
    print(ht.get("Nipun"))
    print(ht.keys())
    

File: hashTables.js
class hashTables{
    constructor(size){
        this.size = new Array(size);
    }

    _hash(str){
        let hashedValue = 0
        if (str && str.length){
            for (let i = 0; i < str.length; i++){
                hashedValue = (hashedValue + str.charCodeAt(i) * i) % this.size.length
                }
        }
        return hashedValue;        
    }

    get(key){
        const location =  this._hash(key);
        if (this.size[location]){
            return this.size[location][1];
        }
        return undefined;
    }

    set(key, value){
        const location = this._hash(key);
        if (!this.size[location]) {
            this.size[location] = [key, value]
        }
        
        return this.size[location];
    }

    get keys(){
        let keys = [];
        for (let i = 0; i < this.size.length; i++){
            if (this.size[i]){
                keys.push(this.size[i][0])
            }
        }
        return keys;
    }
}

var ht = new hashTables(100);
console.log(ht.set("Nakul", "hello"));
console.log(ht.set("Hombre", "weon"));
console.log(ht.get("Nakul"));
console.log(ht.keys);

